### PR TITLE
Importance sampling option for batch_polopt

### PR DESCRIPTION
**Pull request by alexbeloi**
Adding a basic implementation of importance sampling based on [paper from Tang & Abbeel ][1]. 

The main idea is: in between live training iterations, one can use past generated episodes/paths to do additional parameter updates. Importance sampling weights compensate for the bias incurred by using data from past policy versions.

This is implemented primarily in the BatchPolopt class with the aid of a simple sampler (historical_sampler) which applied the importance sampling weights directly to the path data.

Currently works without issue with VPG algorithm. 

When testing NPO based algorithms (e.g. TRPO) on cartpole problem it appears to sometimes runs into nan values in the reward/advantage/return values, ~~I believe it has something to do with the NPO algorithms interaction with previous state_info_vars which may be getting picked up from the importance sampling iterations instead of the live ones.~~ It appears that the log_std parameter becomes NaN at some point when during optimization, lowering the conjugate gradient iterations (cg_iters on TRPO) appears to avoid the NaN's.

[1]: https://papers.nips.cc/paper/3922-on-a-connection-between-importance-sampling-and-the-likelihood-ratio-policy-gradient.pdf

### CODE REVIEW

**Comment by dementrock** [comment]
Hi @alexbeloi, thank you for your contribution!

Currently we'd like to keep the main codebase from having too many configuration parameters. To accommodate this variant, I have refactored the code for `BatchPolopt` to accept a more general sampler object: https://github.com/rllab/rllab/blob/ae95b1aec128ea199caa251405d49da26b68344a/rllab/algos/batch_polopt.py#L185

Could you adjust your code to fit this new interface? This should allow you to pass in your importance-weighted sampler class as a configuration variable when running different algorithms, without needing to further modify the main body of the `BatchPolopt` implementation.

When this is finished, I am happy to merge this under the `contrib/` folder I just created. This could be named e.g. `contrib/alexbeloi/is_sampler` or `contrib/is_sampler` etc. depending on your preference.

Regarding the issue with log_std, I'm a bit surprised that this only occur with NPO but not VPG. An alternative you could try is to increase the `min_std` parameter to something larger: https://github.com/rllab/rllab/blob/ae95b1aec128ea199caa251405d49da26b68344a/rllab/policies/gaussian_mlp_policy.py#L30

**Comment by jachiam** [comment]
Hey, so I looked into the code here, and I don't know if this is quite the source of your error, but it's worth bringing up, because it especially effects the NPO case. You implement the importance sampling weight in the rewards, no? I'm looking at this line:

`path['rewards'] *= is_ratio`

in [your historical sampler](https://github.com/alexbeloi/rllab/blob/importance_sampling/rllab/sampler/historical_sampler.py#L40). 

The `is_ratio` shouldn't be applied directly to the rewards; it should be a weighting factor in forming the sample expectations in the objective and constraints for the various optimization problems in the different algorithms. Particularly, in front of the advantage of a path, and in front of the KL divergence contribution of a path. 

With regards to how this effects the NPO case: the KL divergences in NPO are assumed to be sampled on-policy. But if you only apply the IS weight to the advantages, you haven't accounted for them in the KL divergence constraints; as a result, the KL divergence estimators are no longer on-policy, and you may get strange behavior. 

Hope this helps.

**Comment by alexbeloi** [comment]
Great, I should be able to isolate the changes to a sampler this way.

Regarding the NPO algorithms, I believe Josh Achiam's (missing?) comment regarding the KL divergence is the likely explanation. I'll have to think about if its possible to adjust the computation of the KL divergence on the data level.

**Comment by jachiam** [comment]
Ah, yeah, I deleted it because I thought the best place to put it would have been in an issue ticket on your fork directly, but I couldn't find a way to do that.

Here's the original text of the comment:

Hey, so I looked into the code here, and I don't know if this is quite the source of your error, but it's worth bringing up, because it especially effects the NPO case. You implement the importance sampling weight in the rewards, no? I'm looking at this line:

`path['rewards'] *= is_ratio`

in [your historical sampler](https://github.com/alexbeloi/rllab/blob/importance_sampling/rllab/sampler/historical_sampler.py#L40). 

The `is_ratio` shouldn't be applied directly to the rewards; it should be a weighting factor in forming the sample expectations in the objective and constraints for the various optimization problems in the different algorithms. Particularly, in front of the advantage of a path, and in front of the KL divergence contribution of a path. 

With regards to how this effects the NPO case: the KL divergences in NPO are assumed to be sampled on-policy. But if you only apply the IS weight to the advantages, you haven't accounted for them in the KL divergence constraints; as a result, the KL divergence estimators are no longer on-policy, and you may get strange behavior. 

Hope this helps.

**Comment by alexbeloi** [comment]
All changed completed, apologies for the ugly commit history, let me know if there anything that should be adjusted. There were a couple lines that got cut in your refactoring of batch_polopt.py (top half of part of obtain_samples method cut off) which I re-added, otherwise the changes are isolated to is_sampler.py and vpgis_cartpole.py (example) in the contrib folder .

**Comment by dementrock** [comment]
Hi @alexbeloi, sorry about the lines being cut off. I just pushed a commit that should fix this.

There's a way to fix the commit history. Could you submit a separate pull request which rebase your modified branch from the current master? More details can be seen at https://git-scm.com/book/en/v2/Git-Branching-Rebasing

File: contrib/is_sampler.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            samples = paths[0:n_samples]
+
+        # make duplicate of samples so we don't permanently alter historical data
+        samples = copy.deepcopy(samples)
+
+        dist1 = policy.distribution
+        dist2 = hist_policy.distribution
+        for path in samples:
+            _, agent_infos = policy.get_actions(path['observations'])
+            _, hist_agent_infos = hist_policy.get_actions(path['observations'])
```
Comment (hrtang) [inline contrib/is_sampler.py]: Would you use path["agent_infos"] to retrieve hist_agent_infos instead? That should save some computation on evaluating the policy again.

Comment (alexbeloi) [inline contrib/is_sampler.py]: That's very good, if you use path["agent_infos"] you also don't need to store the old policy at all, just its distribution class.

**Pull request closed without merging**